Mathlib/Tactic/Linter/Multigoal.lean

```diff
@@ -0,0 +1,161 @@
+serve a similar purpose.
+
+TODO:
+* Should the linter flag unnecessary scoping as well?
+  For instance, should
+  ```lean
+  example : True := by
+    · · exact .intro
+  ```
+  raise a warning?
+* Custom support for "accumulating side-goals", so that once they are all in scope
+  they can be solved in bulk via `all_goals` or a similar tactic.
+* In development, `on_goal` has been partly used as a way of silencing the linter
+  precisely to allow the accumulation referred to in the previous item.
+  Maybe revisit usages of `on_goal` and also nested `induction` and `cases`.
+-/
+
+open Lean Elab
+
+namespace Mathlib.Linter
+
+/-- The "multiGoal" linter emits a warning when there are multiple active goals. -/
+register_option linter.style.multiGoal : Bool := {
+  defValue := false
+  descr := "enable the multiGoal linter"
+}
+
+namespace Style.multiGoal
+
+/-- The `SyntaxNodeKind`s in `exclusions` correspond to tactics that the linter allows,
+even though there are multiple active goals.
+Reasons for admitting a kind in `exclusions` include
+* the tactic focuses on one goal, e.g. `·`, `focus`, `on_goal i =>`, ...;
+* the tactic is reordering the goals, e.g. `swap`, `rotate_left`, ...;
+* the tactic is structuring a proof, e.g. `skip`, `<;>`, ...;
+* the tactic is creating new goals, e.g. `constructor`, `cases`, `induction`, ....
+
+Tactic combinators like `repeat` or `try` are a mix of both.
+
+There is some overlap in scope between `ignoreBranch` and `exclusions`.
+-/
+abbrev exclusions : Std.HashSet SyntaxNodeKind := .ofList [
+    -- structuring a proof
+    ``Lean.Parser.Term.cdot,
+    ``cdot,
+    ``cdotTk,
+    ``Lean.Parser.Tactic.tacticSeqBracketed,
+    `«;»,
+    `«<;>»,
+    ``Lean.Parser.Tactic.«tactic_<;>_»,
+    `«{»,
+    `«]»,
+    `null,
+    `then,
+    `else,
+    ``Lean.Parser.Tactic.«tacticNext_=>_»,
+    ``Lean.Parser.Tactic.tacticSeq1Indented,
+    ``Lean.Parser.Tactic.tacticSeq,
+    -- re-ordering goals
+    `Batteries.Tactic.tacticSwap,
+    ``Lean.Parser.Tactic.rotateLeft,
+    ``Lean.Parser.Tactic.rotateRight,
+    ``Lean.Parser.Tactic.skip,
+    `Batteries.Tactic.«tacticOn_goal-_=>_»,
+    `Mathlib.Tactic.«tacticSwap_var__,,»,
+    -- tactic combinators
+    ``Lean.Parser.Tactic.tacticRepeat_,
+    ``Lean.Parser.Tactic.tacticTry_,
+    -- creating new goals
+    ``Lean.Parser.Tactic.paren,
+    ``Lean.Parser.Tactic.case,
+    ``Lean.Parser.Tactic.constructor,
+    `Mathlib.Tactic.tacticAssumption',
+    ``Lean.Parser.Tactic.induction,
+    ``Lean.Parser.Tactic.cases,
+    ``Lean.Parser.Tactic.intros,
+    ``Lean.Parser.Tactic.injections,
+    ``Lean.Parser.Tactic.substVars,
+    `Batteries.Tactic.«tacticPick_goal-_»,
+    ``Lean.Parser.Tactic.case',
+    `«tactic#adaptation_note_»,
+    `tacticSleep_heartbeats_
+  ]
+
+/-- The `SyntaxNodeKind`s in `ignoreBranch` correspond to tactics that disable the linter from
+their first application until the corresponding proof branch is closed.
```

Comment:
Thank you; this is much clearer! I'm stumbling a bit over the term "proof branch" - I don't know a better term though. (Do you?)